### PR TITLE
Allow cache tests to be run from a local branch

### PR DESCRIPTION
**Pull request by brettle**
**Description**
Allow cache tests to be run from a local branch.

**Additional context**
Before this change, the cache tests retrieved the files from GitHub under the branch that the user was in. However, when working on PRs on a local branch, this caused the cache tests to fail because the branch was not yet (and might never be) in the official repo.

### CODE REVIEW

File: tests/cache/cache_environment.py [hunk 1]
```diff
@@ -39,6 +41,18 @@
...
     raise RuntimeError('It was not possible to select a branch name. Running the test suite "cache" group may fail.')
 
+test_url = f'https://raw.githubusercontent.com/cyberbotics/webots/{BRANCH}/README.md'
+try:
+    with urllib.request.urlopen(test_url) as response:
+        pass
+except urllib.error.HTTPError:
+    # This occurs when you are on a branch that does not yet exist in the cyberbotics webots repo (e.g. a local branch).
+    print(f'Unable to access {test_url}.')
+    print(f'Assuming branch {BRANCH} does not exist in the cyberbotics webot repo.')
```
Comment (ad-daniel) [inline tests/cache/cache_environment.py:51]:
```suggestion
    print(f'Assuming branch {BRANCH} does not exist in the cyberbotics Webots repo.')
```

**Review by ad-daniel — CHANGES_REQUESTED** [review]
This should indeed fix the issue when running it locally and doesn't change the test itself, so works for me. @stefaniapedrazzi what do you think?

**Comment by brettle** [comment]
@stefaniapedrazzi can you provide details on your environment and the exceptions you're seeing?

Note that you need remote named "upstream" that refers to the cyberbotics webots repo on GitHub.

**Comment by stefaniapedrazzi** [comment]
I created a local branch named "test_suite_cache" based on `origin/develop` and edited the `resources/branch.txt` file so that I'm sure that the current branch used in the test_suite is the local one (this can also be updated running `scripts/get_git_info/get_git_info.sh`).
This is the common use case where I would expect the test_suite to fail because the corresponding remote branch it not found.

These are the error messages printed in the console:
```
Unable to access https://raw.githubusercontent.com/cyberbotics/webots/test_suite_changes/README.md.
Assuming branch test_suite_changes does not exist in the cyberbotics webot repo.
Traceback (most recent call last):
  File "/home/stefi/webots/tests/cache/cache_environment.py", line 46, in <module>
    with urllib.request.urlopen(test_url) as response:
  File "/usr/lib/python3.10/urllib/request.py", line 216, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.10/urllib/request.py", line 525, in open
    response = meth(req, response)
  File "/usr/lib/python3.10/urllib/request.py", line 634, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.10/urllib/request.py", line 563, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 496, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.10/urllib/request.py", line 643, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/stefi/webots/tests/./test_suite.py", line 34, in <module>
    from cache.cache_environment import update_cache_urls
  File "/home/stefi/webots/tests/cache/cache_environment.py", line 52, in <module>
    BRANCH = subprocess.run(["git", "merge-base", "HEAD", "upstream/develop", "upstream/master"],
  File "/usr/lib/python3.10/subprocess.py", line 524, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['git', 'merge-base', 'HEAD', 'upstream/develop', 'upstream/master']' returned non-zero exit status 128.
```

**Comment by brettle** [comment]
Thanks @stefaniapedrazzi. The issue was that you didn't have a remote named "upstream". I've fixed the code to look for the appropriate remote. Should work for you now.

**Comment by brettle** [comment]
> I'm just wondering if we really need to add the complexity to search for a valid branch or we could simply skip the cache tests.

That would hinder development related to caching on a local branch and for developers who can't create a branch on the official github repo.

**Comment by brettle** [comment]
One of the tests [failed](https://github.com/cyberbotics/webots/actions/runs/4082308814/attempts/1), so I'm rerunning it. I suspect it's an intermittent failure unrelated to this PR.

**Comment by brettle** [comment]
GitHub [seems](https://github.com/cyberbotics/webots/actions/runs/4082308814/jobs/7041334196) to be having some network issues. That would probably explain the previous error as well.

**Comment by ad-daniel** [comment]
On a local branch I'm encountering the same issue reported by Stefania even after the change you made:
```
Unable to access https://raw.githubusercontent.com/cyberbotics/webots/test-branch/README.md.
Assuming branch test-branch does not exist in the github.com/cyberbotics/webots.git repo.
Finding a remote that refers to github.com/cyberbotics/webots.git.
Traceback (most recent call last):
  File "/home/daniel/webots_master/tests/cache/cache_environment.py", line 46, in <module>
    with urllib.request.urlopen(test_url) as response:
  File "/usr/lib/python3.8/urllib/request.py", line 222, in urlopen
    return opener.open(url, data, timeout)
  File "/usr/lib/python3.8/urllib/request.py", line 531, in open
    response = meth(req, response)
  File "/usr/lib/python3.8/urllib/request.py", line 640, in http_response
    response = self.parent.error(
  File "/usr/lib/python3.8/urllib/request.py", line 569, in error
    return self._call_chain(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 502, in _call_chain
    result = func(*args)
  File "/usr/lib/python3.8/urllib/request.py", line 649, in http_error_default
    raise HTTPError(req.full_url, code, msg, hdrs, fp)
urllib.error.HTTPError: HTTP Error 404: Not Found

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "tests/test_suite.py", line 33, in <module>
    from cache.cache_environment import update_cache_urls
  File "/home/daniel/webots_master/tests/cache/cache_environment.py", line 55, in <module>
    remote = next(filter(lambda line: repo in line, remotes.splitlines())).split("\t")[0]
StopIteration
```

**Comment by brettle** [comment]
@ad-daniel, do you not have any remotes? What does `git remote -v` print for you? Also, what OS are you using?

**Comment by ad-daniel** [comment]
I do, perhaps it's related to ssh authentication, the URL is different
```
origin	git@github.com:cyberbotics/webots.git (fetch)
origin	git@github.com:cyberbotics/webots.git (push)
```
I'm on Ubuntu 20

**Comment by brettle** [comment]
Yep. That would do it. Should work now.

**Review by ad-daniel — APPROVED** [review]
It works indeed, thank you!

**Pull request merged**